_posts/2016-10-02-distributed-array-of-commodity-hosting-services.md

```diff
@@ -0,0 +1,184 @@
+enterprise such as ourselves would want an SLA.</q> -- Anonymous Architect</span>
+{: class="look-at-me"}
+
+The above quote shocked me when I was describing to some people how this works.
+These ideas wouldn't seem to be as well-known as I thought, so this post is an
+attempt to do something about that.
+
+This is a straightforward next step from [Google's paper "The Datacenter
+as a computer"][paper]. I'm not sure it has a name. It's an approach I've used to
+reduce hosting costs by a factor of 1,000 for a recent project.
+
+> Our central point is that the datacenters powering many of today’s successful Internet services
+are no longer simply a miscellaneous collection of machines co-located in a facility and wired up
+together. The software running on these systems, such as Gmail or Web search services, execute
+at a scale far beyond a single machine or a single rack: they run on no smaller a unit than clusters
+of hundreds to thousands of individual servers. Therefore, the machine, the computer, is this large
+cluster or aggregation of servers itself and needs to be considered as a single computing unit.
+
+Rather than treating a datacenter as a computer, we can treat any component at
+any level as a composable abstraction.
+
+It came about after discussions with [Jonathan][jcm] and [Adam][adam]. We wanted
+to package commodity services. By doing so, we thought we could [make something
+greater than the sum of the parts](#fn:1)[^1]. It seemed to provide a better
+level of service than any of them would offer in isolation. And so we conceived
+of DACHS[^2].
+
+## Starting point
+
+We have some Platform as a Service (PasS) that we want to run our application
+on. We need to run 100 instances of our application to handle the expected
+load. We'll call this zone A.
+
+The zone A PaaS offers a Service Level Agreement (SLA) of 95% availability[^3].
+
+<table>
+<tr>
+<th>PaaS cost (%)</th>
+<th>Availability (%)</th>
+</tr>
+<tr>
+<td>100</td>
+<td>95%</td>
+</tr>
+</table>
+
+How can we improve on that?
+
+## Add a thing in front of it
+
+Remember the [fundamental theorem of software engineering][ftse]:
+
+> We can solve any problem by adding a layer of indirection
+
+First put in another layer in front of our PaaS. This layer offers an SLA of
+99.99% availability. This layer routes traffic to our stateless application
+running in the PaaS.
+
+<table>
+<tr>
+<th>PaaS cost (%)</th>
+<th>Availability (%)</th>
+</tr>
+<tr>
+<td>100</td>
+<td>94.99</td>
+</tr>
+</table>
+
+This has actually made things worse. The serial availability equation tells us
+that the availability can be calculated from the product of all of the
+component's availability in the network path. So that would be
+`0.9999 * 0.95 = 94.9905%` availability.
+
+We're also now paying for another thing -- the routing layer.
+
+## Cope with single zone failure
+
+We get a similar PaaS, with similar SLAs. We deploy our app into that, and
```

Comment:
Can we say something other than "we get"? It sounds like the result of something we did above, but we're introducing a new zone B here. If it were me, I'd probably say "Let's say we procure a second PaaS, with a similar SLA – we'll call this Zone B." 

This would help me with my subject confusion *

* NB I have had wine.